README.md
1HandZoom
=========
This add-on enables you to zoom in/out by one hand.

This add-on add features below to firefox.

* Edge swipe gesture
    * Swipe up on right edge to zoom out.
    * Swipe down on right edge to zoom in.
* Configurations
  You can change parameters below from about:config.
    * extensions.onehandzoom.gestureSplits
        * Gesture commands are executed when the length of your swipe is more than height_of_your_device/gestureSplits.
        * The valid value range is 2~128. The default value is 32.
    * extensions.onehandzoom.zoomStep
        * The zoom amount of each zoom.
        * The valid value range is 1~100. The default value is 5.


Changelog.txt
* 1.1.0
    - Change how to realize zoom from modifying viewport.zoom to setting
      scale() to style.transform .
    - Enable multiple zoom commands by the single gesture.
* 1.0.2
    - Fix a bug that the view port position is wrong after zoom in/out.
    - Adjust a parameter for sensitivity of a gesture.
* 1.0.1
    - Adjust a parameter that specify the area swipe gesture makes an effect.
* 1.0.0
    - Initial release.


mkpkg.py
#!/usr/bin/env python

import sys
import commands
from xml.etree.ElementTree import *

g_FileList = (
"Changelog.txt",
"Changelog.txt",
"LICENSE.txt",
"README.md",
"bootstrap.js",
"chrome.manifest",
"icon.png",
"icon64.png",
"install.rdf",
"locale"
)

def parseRdf():
    tree = None
    try:
        #register_namespace('em', 'http://www.mozilla.org/2004/em-rdf#')
        tree = parse("install.rdf")
    except Exception as e:
        raise e
    elem = tree.getroot()
    ver = elem.find(".//{http://www.mozilla.org/2004/em-rdf#}version").text
    name = elem.find(".//{http://www.mozilla.org/2004/em-rdf#}name").text
    return ver, name

if __name__ == "__main__":
    ver, name = parseRdf()
    fn = "%s-%s.xpi"%(name, ver)
    print("filename: " + fn)
    fileList = " ".join(g_FileList)
    print fileList
    status, out = commands.getstatusoutput('rm -f %s; \
        grep -rE "DEBUG|TEST" . | grep const; \
        zip -r %s %s'%(fn, fn, fileList))
    print status
    print out


content/preferences.js
const Ci = Components.interfaces;
const Cc = Components.classes;
const Cu = Components.utils;
Cu.import('resource:gre/modules/Services.jsm');

let gStringBundle = null;
let gUserMapK2G = {};

let gBranch = Services.prefs.getBranch('extensions.quickgestures2.');

function log(aMsg) {
    aMsg = 'QuickGestures2#preferences.js: ' + aMsg;
    Services.console.logStringMessage(aMsg);
}

function tr(aName) {
    if (!gStringBundle) {
        let uri = 'chrome://quickgestures2/locale/main.properties';
        gStringBundle = Services.strings.createBundle(uri);
    }

    try {
        return gStringBundle.GetStringFromName(aName);
    } catch (ex) {
        return aName;
    }
}

function sanitize(aString) {
    return aString.replace(/&/g, '&amp;')
                  .replace(/</g, '&lt;')
                  .replace(/>/g, '&gt;')
                  .replace(/"/g, '&quot;');
}

function validateGesture(aGesture, aOverlapCheck) {
    aGesture = aGesture.replace(/U+/g, 'U')
                       .replace(/D+/g, 'D')
                       .replace(/L+/g, 'L')
                       .replace(/R+/g, 'R')
                       .trim();

    if (aGesture && !aGesture.match(/^[UDLR]+$/)) {
        let title = tr('InvalidSeqWarning_Title');
        let msg = tr('InvalidSeqWarning_Msg');
        Services.prompt.alert(null, title, msg);
        throw 'InvalidSeqWarning';
    }

    if (aOverlapCheck) {
        if (aGesture.length == 1){
            let title = tr('OneCharGestureWarning_Title');
            let msg = tr('OneCharGestureWarning_Msg');
            Services.prompt.alert(null, title, msg);
            throw 'OneCharGestureWarning';
        }

        for (let key in gUserMapK2G) {
            let gesture = gUserMapK2G[key];
            if (aGesture && aGesture == gesture) {
                let title = tr('OverlapWarning_Title');
                let msg = tr('OverlapWarning_Msg')
                            .replace('{gesture}', gesture)
                            .replace('{name}', tr(key));

                Services.prompt.alert(null, title, msg);
                throw 'OverlapWarning';
            }
        }
    }
    return aGesture;
}

function showRemapDialog() {
    let target = this;
    let name = target.getAttribute('qg-name');
    let key = target.getAttribute('qg-key');
    let gesture;

    try {
        gesture = validateGesture(target.textContent, false);
    } catch(ex) {
        gesture = '';
    }

    // Display prompt to remap gesture
    let title = name; // tr('RemapPrompt_Title');
    let msg = tr('RemapPrompt_Msg');
    let ret = {value: gesture};
    let ok = Services.prompt.prompt(null, title, msg, ret, null, {});
    if (ok) {
        let newGesture = ret.value;

        if (newGesture == gesture)
            return;

        try {
            newGesture = validateGesture(newGesture, true);
        } catch(ex) {
            return;
        }

        while (target.firstChild)
            target.removeChild(target.firstChild);

        let textNode = target.ownerDocument.createTextNode(newGesture || '  ');  
        target.appendChild(textNode);

        gUserMapK2G[key] = newGesture;
        gBranch.setCharPref('mapk2g', JSON.stringify(gUserMapK2G));
    }
}

function showResetGesturesDialog() {
    let title = tr('ResetGesturesDialog_Title');
    let msg = tr('ResetGesturesDialog_Msg');

    let ok = Services.prompt.confirm(null, title, msg);
    if (ok) {
        gBranch.clearUserPref('mapk2g');
        window.location.reload();
    }
}

function showResetSettingsDialog() {
    let title = tr('ResetSettingsDialog_Title');
    let msg = tr('ResetSettingsDialog_Msg');

    let ok = Services.prompt.confirm(null, title, msg);
    if (ok) {
        gBranch.clearUserPref('layer.visible');
        gBranch.clearUserPref('toast.visible');
        gBranch.clearUserPref('threshold.splits');
        gBranch.clearUserPref('threshold.angle');
        gBranch.clearUserPref('threshold.timeout');
        window.location.reload();
    }
}


function onLoad() {
    // Gestures Panel
    let mapk2g_json = gBranch.getCharPref('mapk2g');
    
    gUserMapK2G = JSON.parse(mapk2g_json);

    let listItemTemplate =
        '<div class="content pref-gesture">\n' +
        '  <div class="g-cmd">{name}</div>\n' +
        '  <div class="g-seq" id="g-{key}" qg-key="{key}" qg-name="{name}" >{gesture}</div>\n' +
        '</div>\n';

    let listItems = '';

    for (let commandKey in gUserMapK2G) {
        let gesture = gUserMapK2G[commandKey];
        let name = sanitize(tr(commandKey));

        let listItem = listItemTemplate.replace(/{gesture}/g, gesture || '&#160;')
                                       .replace(/{name}/g, name)
                                       .replace(/{key}/g, commandKey);
        listItems += listItem;
    }

    // Insert items
    document.getElementById('gestures').innerHTML = listItems;

    for (let item of document.getElementsByClassName('g-seq')) {
        item.addEventListener('click', showRemapDialog, false);
    }

    // Reset button
    document.getElementById('gestures-reset-button').addEventListener('click', showResetGesturesDialog, false);
    document.getElementById('settings-reset-button').addEventListener('click', showResetSettingsDialog, false);


    // Settings Panel
    // Set default values
    document.getElementById('layerVisible').checked = gBranch.getBoolPref('layer.visible');
    document.getElementById('toastVisible').checked = gBranch.getBoolPref('toast.visible');
    document.getElementById('splitsSeekbarInput').value = gBranch.getIntPref('threshold.splits');
    document.getElementById('angleSeekbarInput').value = gBranch.getIntPref('threshold.angle');

    let timeout = gBranch.getIntPref('threshold.timeout');
    document.getElementById('timeoutSeekbarInput').value = Math.floor(100 * Math.LOG10E * Math.log(2 * timeout));

    document.getElementById('splitsSeekbarOutput').value = gBranch.getIntPref('threshold.splits');
    document.getElementById('angleSeekbarOutput').value = gBranch.getIntPref('threshold.angle');
    document.getElementById('timeoutSeekbarOutput').value = gBranch.getIntPref('threshold.timeout') + 'ms';


    // Set event listeners
    document.querySelector('.pref-checkbox[for="layerVisible"]').addEventListener('click', function() {
        let cb = document.getElementById('layerVisible');
        cb.checked = !cb.checked;
        gBranch.setBoolPref('layer.visible', cb.checked);
    }, false);

    document.querySelector('.pref-checkbox[for="toastVisible"]').addEventListener('click', function() {
        let cb = document.getElementById('toastVisible');
        cb.checked = !cb.checked;
        gBranch.setBoolPref('toast.visible', cb.checked);
    }, false);

    document.getElementById('timeoutSeekbarInput').addEventListener('input', function() {
        let timeout = Math.floor(Math.pow(10.0, parseFloat(this.value)/100-0.301));
        document.getElementById('timeoutSeekbarOutput').value = timeout + 'ms';
        gBranch.setIntPref('threshold.timeout', timeout);
    }, false);

    document.getElementById('splitsSeekbarInput').addEventListener('input', function() {
        document.getElementById('splitsSeekbarOutput').value = this.value;
        gBranch.setIntPref('threshold.splits', this.value);
    }, false);

    document.getElementById('angleSeekbarInput').addEventListener('input', function() {
        document.getElementById('angleSeekbarOutput').value = this.value;
        gBranch.setIntPref('threshold.angle', this.value);
    }, false);


    // Disable the seekbars because the new `range` input
    // will be supported from Firefox 23.
    //
    let firefoxApp = Cc['@mozilla.org/xre/app-info;1'].getService(Ci.nsIXULAppInfo);
    if (parseFloat(firefoxApp.version) < 23.0) {
        document.getElementById('timeout-section').className += ' disabled';
        document.getElementById('recognition-section').className += ' disabled';
    }
}

document.addEventListener('DOMContentLoaded', onLoad, false);



bootstrap.js
// vim: set sw=4 expandtab:
const Ci = Components.interfaces;
const Cu = Components.utils;
Cu.import("resource://gre/modules/Services.jsm");

const DEBUG = false; // If false, the debug() function does nothing.
//const DEBUG = true; // If false, the debug() function does nothing.
const SPLIT_MIN = 2;
const SPLIT_MAX = 128;
const ZOOM_STEP_MIN = 1;
const ZOOM_STEP_MAX = 100;
const GESTURE_VALID_RANGE_RATIO = 0.9;
const ZOOM_IN_MAX = 10.0;
const ZOOM_OUT_MIN = 0.5;

let windowListener = {
    onOpenWindow: function(aWindow) {
        let win = aWindow.QueryInterface(Ci.nsIInterfaceRequestor)
                         .getInterface(Ci.nsIDOMWindowInternal
                                                || Ci.nsIDOMWindow);

        win.addEventListener('UIReady', function() {
            win.removeEventListener('UIReady', arguments.callee, false);
            OneHandZoom.load(win);
        }, false);
    },

    // Unused
    onCloseWindow: function(aWindow) {},
    onWindowTitleChange: function(aWindow, aTitle) {},
};

//===========================================
// OneHandZoom
//===========================================
let OneHandZoom = {
    install: function() {
        debug('install()');
    },

    uninstall: function() {
        debug('uninstall()');
    },

    _setupDefaultPrefs: function() {
        debug('_setupDefaultPrefs()');

        let branch = Services.prefs.getDefaultBranch('extensions.onehandzoom.');
        branch.setIntPref('gestureSplits', 32);  // Division number of screen width (2 ~ 16)
        branch.setIntPref('zoomStep', 5);  // actual_zoom_step = zoomStep/100
    },

    _updateThresholdSub: function(name, min, max) {
        let val = this._branch.getIntPref(name);
        if (val < min || val > max) {
            val = (val < min) ? min : max;
            this._changePref = true;
            this._branch.setIntPref(name, val);
        }
    },

    _updateThreshold: function() {
        this._updateThresholdSub('gestureSplits', SPLIT_MIN, SPLIT_MAX);
        this._updateThresholdSub('zoomStep', ZOOM_STEP_MIN, ZOOM_STEP_MAX);
        this._splits = this._branch.getIntPref('gestureSplits');
        this._zoomStep = this._branch.getIntPref('zoomStep')/100.0;
    },

    init: function() {
        debug('init()');

        this._branch = null;
        this._setupDefaultPrefs();

        if (!this._branch) {
            this._branch = Services.prefs.getBranch('extensions.onehandzoom.');
            this._updateThreshold();
            this._branch.addObserver('', this, false);
        }
    },

    uninit: function() {
        debug('uninit()');

        if (this._branch) {
            this._branch.removeObserver('', this);
            this._branch = null;
        }
    },

    _menuIdZoomIn: null,
    _menuIdZoomOut: null,

    load: function(aWindow) {
        debug('load(' + aWindow + ')');

        if (!aWindow)
            return;

        let deck = aWindow.BrowserApp.deck;
        deck.addEventListener('load', this, true);
        deck.addEventListener('touchstart', this, true);
        deck.addEventListener('touchmove', this, true);
        deck.addEventListener('touchend', this, true);

        // add menu
        //this._menuIdZoomIn = aWindow.NativeWindow.menu.add("Zoom In", null, this._cbZoomIn);
        //this._menuIdZoomOut = aWindow.NativeWindow.menu.add("Zoom Out", null, this._cbZoomOut);
    },

    /*
    _cbZoomIn: function() {
        zoomIn(this._zoomStep));
    },
    _cbZoomOut: function() {
        zoomOut(this._zoomStep));
    },
    */

    unload: function(aWindow) {
        debug('unload(' + aWindow + ')');

        if (!aWindow)
            return;

        //aWindow.NativeWindow.menu.remove(this._menuIdZoomIn);
        //aWindow.NativeWindow.menu.remove(this._menuIdZoomOut);

        let deck = aWindow.BrowserApp.deck;
        deck.removeEventListener('load', this, true);
        deck.removeEventListener('touchstart', this, true);
        deck.removeEventListener('touchmove', this, true);
        deck.removeEventListener('touchend', this, true);
    },

    observe: function(aSubject, aTopic, aData) {
        debug('observe(' + aSubject + ', ' + aTopic + ', ' + aData + ')');

        if (this._changePref) {
            this._changePref = false;
            return;
        }

        switch (aData) {
        case 'gestureSplits':
        case 'zoomStep':
            this._updateThreshold();
            break;
        }
    },

    handleEvent: function(aEvent) {
        let fingers = null;
        let tab = getSelectedTab();

        switch (aEvent.type) {
        case 'load':
            // clear current zoom status
            tab._currentZoom = 1.0;
            break;
        case 'touchstart':
            this._startStroke(aEvent);
            break;
        case 'touchmove':
            this._progressStroke(aEvent);
            break;
        case 'touchend':
            fingers = aEvent.touches.length;
            if (fingers != 0) {
                break;
            }
            this._stopStroke(aEvent);
            break;
        }
    },

    _startStroke: function(aEvent) {
        debug('_startStroke(' + aEvent + ')');
        const touch = aEvent.touches.item(0);
        this._chromeWindow = getChromeWindow();
        this._selectedTab = this._chromeWindow.BrowserApp.selectedTab;
        this._contentWindow = this._selectedTab.browser.contentWindow;

        const range = this._contentWindow.innerWidth * GESTURE_VALID_RANGE_RATIO;
        if(touch.screenX < range) {
            // touches in this area is ignored
            this._cancelStroke(aEvent);
            return;
        }

        const viewport = this._selectedTab.getViewport();
        const cssh = viewport.cssHeight;

        this._threshold = Math.round(cssh/this._splits);

        this._startY = touch.screenY;
        this._inProgress = true;
    },

    _progressStroke: function(aEvent) {
        const touch = aEvent.touches.item(0);
        const contentWindow = this._contentWindow;
        const innerWidth = contentWindow.innerWidth;
        const range = innerWidth * GESTURE_VALID_RANGE_RATIO;
        if(touch.screenX < range) {
            // touches in this area is ignored
            this._cancelStroke(aEvent);
            return;
        }
        aEvent.preventDefault(); // discard the event

        if (!this._inProgress)
            return;

        const x = touch.screenX;
        const y = touch.screenY;
        const viewport = this._selectedTab.getViewport();
        const cssh = viewport.cssHeight;
        let w = this._selectedTab.window.document.documentElement.clientWidth;
        let h = this._selectedTab.window.document.documentElement.clientHeight;
        debug("x, y: w, h: " + x + "," + y + "," + w + "," + h + "," + contentWindow.innerHeight + ", " + (y/h) + "," + (y/cssh));

        let diff = this._startY - y;
        debug("diff: " + diff + "," + this._threshold);
        if(diff < -this._threshold) {
            zoomIn(this._zoomStep);
            this._startY = y;
        }
        else if(diff > this._threshold) {
            zoomOut(this._zoomStep);
            this._startY = y;
        }
    },

    _cancelStroke: function(aEvent) {
        debug('_cancelStroke(' + aEvent + ')');
        this._inProgress = false;
        this._stopStroke(aEvent);
    },

    _stopStroke: function(aEvent) {
        debug('_stopStroke(' + aEvent + ')');
        this._inProgress = false;
    },
};

//===========================================
// bootstrap.js API
//===========================================
function install(aData, aReason) {
    OneHandZoom.install();
}

function uninstall(aData, aReason) {
    if (aReason == ADDON_UNINSTALL) {
        OneHandZoom.uninstall();
    }
}

function startup(aData, aReason) {
    // General setup
    OneHandZoom.init();

    // Load into any existing windows
    let windows = Services.wm.getEnumerator('navigator:browser');
    while (windows.hasMoreElements()) {
        let win = windows.getNext().QueryInterface(Ci.nsIDOMWindow);
        if (win) {
            OneHandZoom.load(win);
        }
    }

    // Load into any new windows
    Services.wm.addListener(windowListener);
}

function shutdown(aData, aReason) {
    // When the application is shutting down we normally don't have to clean
    // up any UI changes made
    if (aReason == APP_SHUTDOWN) {
        return;
    }

    // Stop listening for new windows
    Services.wm.removeListener(windowListener);

    // Unload from any existing windows
    let windows = Services.wm.getEnumerator('navigator:browser');
    while (windows.hasMoreElements()) {
        let win = windows.getNext().QueryInterface(Ci.nsIDOMWindow);
        if (win) {
            OneHandZoom.unload(win);
        }
    }

    // General teardown
    OneHandZoom.uninit();
}

//===========================================
// Utilities
//===========================================
function debug(aMsg) {
    if (!DEBUG) {
        return;
    }
    aMsg = 'OneHandZoom: ' + aMsg;
    console.log(aMsg);
}

function getChromeWindow() {
    return Services.wm.getMostRecentWindow('navigator:browser');
}

function getSelectedTab() {
    let selectedTab = getChromeWindow().BrowserApp.selectedTab;
    return selectedTab;
}

function initZoomIfNull() {
    let selectedTab = getSelectedTab();
    if(selectedTab._currentZoom == null) {
        selectedTab._currentZoom = 1.0;
    }
}

function zoomIn(step) {
    let selectedTab = getSelectedTab();
    let chromeWindow = getChromeWindow();

    initZoomIfNull();

    selectedTab._currentZoom *= (1.0 + step);
    if(selectedTab._currentZoom > ZOOM_IN_MAX) {
        selectedTab._currentZoom = ZOOM_IN_MAX;
    }
    let doc  = chromeWindow.content.document;
    doc.body.style.transformOrigin='left top';
    doc.body.style.transform='scale(' + selectedTab._currentZoom + ')';

    debug("=== zoomIn: currentZoom: " + selectedTab._currentZoom);
}

function zoomOut(step) {
    let selectedTab = getSelectedTab();
    let chromeWindow = getChromeWindow();

    initZoomIfNull();

    selectedTab._currentZoom /= (1.0 + step);
    if(selectedTab._currentZoom < ZOOM_OUT_MIN) {
        selectedTab._currentZoom = ZOOM_OUT_MIN;
    }
    let doc  = chromeWindow.content.document;
    doc.body.style.transformOrigin='left top';
    doc.body.style.transform='scale(' + selectedTab._currentZoom + ')';

    debug("=== zoomOut: currentZoom: " + selectedTab._currentZoom);
}

